AWS Batch Executor

syedahsn opened the pull request:
Overview
------------
This PR introduces the AWS Batch Executor. This Executor can be configured to run Airflow tasks using AWS Batch. It is based on an initial contribution from @aelzeiny.

From the README:
```
This is an Airflow executor powered by Amazon Batch. Each task scheduled by Airflow is run inside a
separate container, scheduled by Batch. Some benefits of an executor like this include:

1. Scalability and Lower Costs: AWS Batch allows the ability to dynamically provision the resources needed to execute tasks.
   Depending on the resources allocated, AWS Batch can autoscale up or down based on the workload,
   ensuring efficient resource utilization and reducing costs.
2. Job Queues and Priority: AWS Batch provides the concept of job queues, allowing
   the ability to prioritize and manage the execution of tasks. This ensures that when multiple
   tasks are scheduled simultaneously, they are executed in the desired order of priority.
3. Flexibility: AWS Batch supports Fargate (ECS), EC2 and EKS compute environments. This range of
   compute environments, as well as the ability to finely define the resources allocated to
   the compute environments gives a lot of flexibility to users in choosing the most suitable
   execution environment for their workloads.
4. Rapid Task Execution: By maintaining an active worker within AWS Batch, tasks submitted to
   the service can be executed swiftly. With a ready-to-go worker, there's minimal startup delay,
   ensuring tasks commence immediately upon submission. This feature is particularly advantageous
   for time-sensitive workloads or applications requiring near-real-time processing,
   enhancing overall workflow efficiency and responsiveness.
```
This PR comes with *most* features now included in the ECS Executor, but we will continue to update both executors as we work to improve them. 

Review Notes
------------------
Similar to the ECS Executor, this PR comes as a fully functional feature, which is great for testing, but it does mean that the PR is large. When reviewing, it isn't necessary to go through every single line. Instead, it would be more beneficial to have read through the documentation, and glance at how the Executor is implemented. 

There are a lot of similarities in the code between the Batch Executor and the ECS Executor - this is by design. As we continue to refine the process of writing custom Executors, we are converging on an optimal framework to write efficient, reliable, and fault-tolerant custom Executors. 

Testing
---------
There is extensive unit testing which has near 100% line coverage in most cases:
![BatchExecutorCodeCoverage](https://github.com/apache/airflow/assets/103602455/71fc531f-c2e1-4592-b2a3-15eb3684e60f)


<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

aelzeiny committed:
Introduce AWS Executors to the Amazon provider package. The EcsFargateExecutor and BatchExecutor launches Airflow tasks on AWS ECS/Fargate service and the AWS Batch service, respectively.

syedahsn committed:
Initial commit to break up code into different modules

syedahsn committed:
WIP: Changes to get executor working

syedahsn committed:
Added option to allow users to pass in config options.

syedahsn committed:
prevent any failures in the executor from killing the scheduler

syedahsn committed:
WIP: Move add_job inside try block

syedahsn committed:
Add documentation for Batch Executor
Rebase from main

syedahsn committed:
Health check and Token Expiration

syedahsn committed:
Retries on task failures

syedahsn committed:
Remove old config files

syedahsn committed:
Add missing doc string, and update outdated doc string

syedahsn committed:
Add unit tests, and move Dockerfile to executor folder

syedahsn committed:
Fix failing tests

syedahsn committed:
Add doc string to base config key class

Taragolis commented on airflow/providers/amazon/aws/executors/batch/utils.py:
```diff
@@ -0,0 +1,157 @@
+if TYPE_CHECKING:
+    from airflow.models.taskinstance import TaskInstanceKey
+from airflow.providers.amazon.aws.executors.utils.base_config_keys import BaseConfigKeys
+from airflow.utils.state import State
```
```suggestion
from airflow.providers.amazon.aws.executors.utils.base_config_keys import BaseConfigKeys
from airflow.utils.state import State

if TYPE_CHECKING:
    from airflow.models.taskinstance import TaskInstanceKey
```

Taragolis commented on airflow/providers/amazon/aws/executors/batch/batch_executor_config.py:
```diff
@@ -0,0 +1,78 @@
+        raise KeyError("Eks jobs are not currently supported.")
+
+    # some checks with some helpful errors
+    assert isinstance(job_kwargs, dict)
```
We should avoid to use assert outside of the tests: https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#don-t-use-asserts-outside-tests

If it for make mypy happy, then it fine, just need keep it under the TYPE_CHECKING block

```suggestion
    if TYPE_CHECKING:
        assert isinstance(job_kwargs, dict)
```

if it intends to check in runtime then better raise appropriate error, e.g. TypeError

syedahsn replied:
That's a good catch! I've updated it to do the check without using an assert statement, which is how its done in the ECS Executor. Thanks for reviewing!

o-nikolas replied:
> I've updated it to do the check without using an assert statement

I still see an assert?

Taragolis reviewed (commented):
Minor things that I've found

syedahsn committed:
Remove assert statement from non testing code
Fix import statements order

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+    """
+    The Airflow Scheduler creates a shell command, and passes it to the executor.
+
+    This Batch Executor simply runs said airflow command a resource provisioned and managed
```
```suggestion
    This Batch Executor simply runs said airflow command in a resource provisioned and managed
```

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+    This Batch Executor simply runs said airflow command a resource provisioned and managed
+    by AWS Batch. It then periodically checks in with the launched jobs (via job-ids) to
+    determine the status.
+    The `submit_job_kwargs` configuration points to a dictionary that returns a dictionary. The
```
This phrasing is strange?

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+        for job in describe_job_response:
+            if job.get_job_state() == State.FAILED:
+                self._handle_failed_job(job)
+            elif job.get_job_state() == State.SUCCESS:
```
Are there any other possible states that need handling? Should there be an `else` here to catch those or future states?

syedahsn replied:
There shouldn't be any other relevant states. `get_job_state` can return `State.QUEUED`, `State.RUNNING`, `State.FAILED`, or `State.SUCCESS`. `State.QUEUED` is kind of a catchall for various intermediate states. For both the QUEUED and RUNNING states, we don't want to do anything, so that only leaves the 2 states handled above.

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+                task_key = self.active_workers.pop_by_id(job.job_id)
+                self.success(task_key)
+
+    def _handle_failed_job(self, job):
```
Maybe I have this wrong, but this is being called from `sync_running_jobs` which is just getting the latest status from already successfully submitted jobs, yet the logic below is assuming it's an API failure submitting the job (comparing against max submit attempts).

If the job/task legitimately failed while running (some issue with the user code in Airflow) shouldn't we not be retrying it here? We should leave that up to the Airflow scheduler (if the user set retries on their task/dag).

syedahsn replied:
A failed job here means a job that AWS Batch determines has failed. This happens in cases where the container failed to start, or some other issue which caused Batch to determine that the job failed. This is not the same as an Airflow failed task, which fails if Airflow determines that the DAG failed in some way. It is possible for Batch to declare a job successful, and Airflow to determine the task than run in that job failed. 
Here, we are dealing with the case the something prevented Batch from launching the container that should run the Airflow process. 

I will add a comment in the function to explain what the function is doing.

o-nikolas replied:
Sounds good, thanks for the clarification on that behaviour. It is indeed counter intuitive!

o-nikolas replied:
I don't see any explanatory comment here.

syedahsn replied:
Added now.

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+                    self.fail(key=key)
+                else:
+                    batch_job.attempt_number += 1
+                    self.pending_jobs.appendleft(batch_job)
```
If a task fails to submit, I think we should probably wait a little bit before retrying it, moving on to the next task in the list in the meantime.

syedahsn replied:
I've updated this to use the same mechanism used in the ECS Executor. It now appends to the right of the queue, and determines a `next_attempt_time` which determines when the task will be attempted next.

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+                self.log.error("Batch DescribeJobs API Response: %s", boto_describe_tasks)
+                raise BatchExecutorException(
+                    f"DescribeJobs API call does not match expected JSON shape. Are you sure that the correct version of Boto3 is installed? {err}"
+                )
```
I wonder if we need all of this? Can we not just let boto raise the validation error?

syedahsn replied:
You're correct, this is somewhat redundant. I'll remove it.

o-nikolas commented on airflow/providers/amazon/aws/executors/batch/batch_executor.py:
```diff
@@ -0,0 +1,386 @@
+        submit_job_api = deepcopy(self.submit_job_kwargs)
+        submit_job_api["containerOverrides"].update(exec_config)
+        submit_job_api["containerOverrides"]["command"] = cmd
+        return submit_job_api
```
Shouldn't you follow the update strategy that we settled on for the ECS executor? I.e. update the whole dict instead of just the container overrides?

syedahsn replied:
Yep, I missed this. I've added it now. Thanks!

o-nikolas commented on docs/apache-airflow-providers-amazon/executors/batch-executor.rst:
```diff
@@ -0,0 +1,318 @@
+   :end-before: .. END CONFIG_OPTIONS_PRECEDENCE
+
+.. note::
+   ``exec_config`` is an optional parameter that can be provided to operators. It is a dictionary type and in the context of the Batch Executor it represents a ``submit_job_kwargs`` configuration which is then updated over-top of the ``submit_job_kwargs`` specified in Airflow config above (if present). It is a recursive update which essentially applies Python update to each nested dictionary in t ...
```
Your code doesn't yet do this recursive update, I left a comment about it above.

o-nikolas commented on docs/apache-airflow-providers-amazon/executors/batch-executor.rst:
```diff
@@ -0,0 +1,318 @@
+
+There are 3 steps involved in getting a Batch Executor to work in Apache Airflow:
+
+1. Creating a database that Airflow and the tasks running executed by Batch can connect to.
```
```suggestion
1. Creating a database that Airflow and the tasks executed by Batch can connect to.
```

o-nikolas commented on docs/apache-airflow-providers-amazon/executors/batch-executor.rst:
```diff
@@ -0,0 +1,318 @@
+Configure Airflow
+~~~~~~~~~~~~~~~~~
+
+To configure Airflow to utilize the Batch Executor and leverage the resources we've set up, create a script (e.g., ``batch_executor_config.sh``) with the following contents:
```
I wouldn't suggest the script thing, that's a pretty specific approach. The important thing to convey is that these Airflow configuration changes need to be made (in airflow.cfg, env vars, etc)

o-nikolas reviewed (commented):
I'm halfway through, but need to go for lunch, so I'm "committing" the first batch (no pun intended) of comments :)

Looking great so far!

o-nikolas commented on docs/apache-airflow-providers-amazon/executors/general.rst:
```diff
@@ -0,0 +1,331 @@
+Airflow environment. It also includes options to load DAGs (Directed
+Acyclic Graphs) from either an S3 bucket or a local folder.
+
+Download this image to use for the docker build commands below or create
```
Download the image? Do we offer it anywhere for download?

o-nikolas commented on tests/providers/amazon/aws/executors/batch/test_batch_executor.py:
```diff
@@ -0,0 +1,586 @@
+        assert running_job.get_job_state() == State.RUNNING
+
+    def test_success_jobs(self):
+        """Jobs that have been launched are identified as 'running'"""
```
Here and below: copy/paste doc strings

syedahsn committed:
Make requested changes
 -Apply recursive updates to exec_config parameter
 -Apply waits between each task retry attempt
 -Unit tests

syedahsn committed:
Fix test name
Add doc string to functions

syedahsn committed:
Fix failing test

Taragolis reviewed (approved):
Looking good to me, if it take in account that executor has an experimental status

o-nikolas reviewed (approved):
Some minor comments on old threads.

Can you add one last change to the docs to list this as an experimental executor (as we currently do for ECS [here](docs/apache-airflow-providers-amazon/executors/index.rst:27:)).

syedahsn committed:
Add comment about Batch Job failure
Remove assert statement
